File: princeton-part-1/leet_code/medium/three_sum.cpp
#include<iostream>
#include<vector>
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
      vector<vector<int>> triplets;
      sort(nums.begin(), nums.end());

      set<vector<int>> set_of_vectors;
      vector<int>triplet;
      
      for(int i=0; i<nums.size(); i++){
	int k = i+1;
	int j = nums.size()-1;
	while(k<j){
	  if(nums[i]+nums[k] == -nums[j]){
	    if(triplet.size()==0){
	      triplet.push_back(nums[i]);
	      triplet.push_back(nums[k]);
	      triplet.push_back(nums[j]);
	    }
	    else{
	      triplet[0] =  nums[i];
	      triplet[1] =  nums[k];
	      triplet[2] =  nums[j];
	    }

	    if(set_of_vectors.count(triplet)==0){
	      set_of_vectors.insert(triplet);
	    }

	    k++;
	    j--;
	  }

	  else if(nums[i]+nums[k] < -nums[j]){
	    k++;
	  }
	  else{
	    j--;
	  }
	}
      }

      for (auto it = set_of_vectors.begin(); it != set_of_vectors.end(); it++) {
        triplets.push_back(*it);
      } 
      
      return triplets;
    }
};

int main(){
  Solution soln;
  vector<int> nums = {-1,0,1,2,-1,-4};
  vector <vector <int>> triplets = soln.threeSum(nums);
  for(int i=0; i<triplets.size(); i++){
    cout << triplets[i][0] << ", " << triplets[i][1] << ", " << triplets[i][2] << "\n";
  }
}

File: princeton-part-1/leet_code/medium/three_sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        ans = {}
        n = len(nums)
        nums.sort()
        
        for k in range(n):
            i = k+1
            j = n-1
            
            while(i<j):
                if nums[i] + nums[j] == -nums[k] and (nums[k],nums[i],nums[j]) not in ans: 
                    ans[(nums[k],nums[i],nums[j])] = 1
                    i+=1
                elif nums[i] + nums[j] < -nums[k] : 
                    i+=1
                else:
                    j-=1
        return [list(k) for k in ans]

